Added some accessor functions in `Component`

joshuahansel opened the pull request:
These changes are a step toward completion of #22354.

The changes are as follows:

- Created `Component::constMesh()`. This API is used to get a constant reference to THMMesh in places where the THMMesh should not be modified. This is to be extra careful not to modify THMMesh after mesh setup, since in the future, the ultimate mesh (combining THMMesh with mesh generators) will not be able to get these changes, since there will be a copy of THMMesh made at the point that it enters the MeshGenerator system.
- Modified `Component::mesh()` (the non-const version) to check that it is not being called after mesh setup is complete.
- Made `Component::getTHMProblem()` accessor function. The THMProblem reference member variable in `Component` (named `_sim`) will need to be deleted because THMProblem will no longer be available at the time of component construction. Thus the reference will be encapsulated to allow some other implementation.
- Used new accessor functions in all components.

After this is merged, THM-based applications will need to be similarly updated. Then `_mesh` can be made a private member variable in `Component`, and `_sim` can be removed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1241017/) on 5fb9555 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3b891b3ab8d6dc8b986fecff9f9dd16b71c93b55`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1241305/) on a32062e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3b891b3ab8d6dc8b986fecff9f9dd16b71c93b55`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1241365/) on 68d1349 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22773/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1241378/) on 68d1349 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3b891b</th>
      <th colspan=4>#22773 68d134</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.39%</td>
      <td>84.39%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80649</td>
      <td>80648</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14916</td>
      <td>14917</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22773/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22773/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>3b891b</th>
      <th colspan=4>#22773 68d134</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.18%</td>
      <td>91.19%</td>
      <td>+0.01%</td>
      <td>97.12%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12370</td>
      <td>12401</td>
      <td>+31</td>
      <td>303</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1197</td>
      <td>1198</td>
      <td>+1</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22773/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22773/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22773/coverage/xfem)

</details>



This comment will be updated on new commits.

licharlot commented on modules/thermal_hydraulics/include/components/Component.h:
```diff
@@ -223,6 +223,16 @@ class Component : public THMObject, public LoggingInterface, public NamingInterf
+  /**
+   * Gets a const reference to the THM problem
+   */
+  const THMProblem & getTHMProblem() const;
```
So here you can overload, but not for mesh?

joshuahansel replied:
Note that this particular overload is unnecessary/not useful, and I have since deleted it, since both `const THMProblem &` and `THMProblem &` versions can be `const` methods. In cases where an accessor method is returning a reference to data that actually belongs to the class, there should be the two different versions, but here, the member variable is a reference itself, but the data lives somewhere else.

As far as why the mesh accessor function cannot be overloaded: It *can* be overloaded, but when you overload, you lose the ability to be explicit about which version to use - the compiler decides based on the context. If the object is `const`, it chooses the `const` method; else, it chooses the other. This case is a bit special because the two versions are actually different - one makes a check, and the other doesn't.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1243435/) on 50e3b4b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22773/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e842bc522b5f93d6a6e790ff9ddd8179eb358fac`

joshuahansel committed:
Created const mesh getter function in Component

Added const and non-const mesh getter functions in Component.
Note that the const version could not overload the other due
to the compiler choosing the non-const version in non-const
member functions.

The non-const version checks that mesh setup has not been completed
and throws an error if it has. There should not be any mesh modification
past that point.

Component::_mesh still needs to be made private, but applications
need to be patched to use the new interfaces first.

Refs #22354

joshuahansel committed:
Made getter functions for THM problem in Component

Components will no longer be able to have a member function
reference to THMProblem because Components will be constructed
before the creation of THMProblem. Thus getter functions are
created to encapsulate this reference. Applications will then
be updated to use the getters instead of the reference. Then
the reference will be removed at that point.

Refs #22354
